ENH Add parameter in colorbar in plot_confusion_matrix

avigupta2612 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Added feature to make colorbar optional as mentioned in #16519 
#### What does this implement/fix? Explain your changes.

- Added an argument colorbar in plot_confusion_matrix function and Set the colorbar to True by Default
- Made same changes to the Confusion matrix display class's plot function and changed
'
fig.colorbar(self.im_, ax=ax)
'
to 
'
if self.colorbar:
            fig.colorbar(self.im_, ax=ax)
'
#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

avigupta2612 committed:
optional colorbar

avigupta2612 commented:
Can anyone tell what what went wrong in this and how can I correct this?

cmarmo commented:
Hi @avigupta2612, your code have a linting issue.
scikit-learn use lint `flake8` directives:  in order to check your code locally you can run
```
$ flake8 <your_modified_file>.py
```
Hope this can help.

avigupta2612 committed:
lint edit

avigupta2612 commented:
So this is my first PR. Plot_confusion_matrix tests are failing. So, do I need to edit the tests? If so, how can I do it?

cmarmo commented:
Tests related to each module of scikit-learn are in `sklearn/<name_of_the_module>/tests/` directory. The build logs (e.g. [this one](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=16552&view=logs&j=78a0bf4f-79e5-5387-94ec-13e67d216d6e&t=f1857171-4a53-55c7-3ab5-90acfe091baa)) tell you which test is failing and the specific error.
In particular, you are in the process of adding an argument to `plot_confusion_matrix`, you should then check where in the tests this argument needs to be explicitly added.

avigupta2612 committed:
test edit

cmarmo commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -70,6 +71,7 @@ def plot(self, *, include_values=True, cmap='viridis',
             created.
 
+        colorbar : matplotlib colorbar, default=True
         Returns
```
If you check the [rendering of the documentation](https://102981-843222-gh.circle-artifacts.com/0/doc/modules/generated/sklearn.metrics.ConfusionMatrixDisplay.html#sklearn.metrics.ConfusionMatrixDisplay.plot) you see that you need a blank line before the "Returns" paragraph in order to keep the correct indentation.

cmarmo commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -70,6 +71,7 @@ def plot(self, *, include_values=True, cmap='viridis',
             Axes object to plot on. If `None`, a new figure and axes is
             created.
 
+        colorbar : matplotlib colorbar, default=True
```
You are adding a new parameter, a useful resource is the [scikit-learn contributing documentation](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation), in order to standardize the parameter description.
Also, could you please add a `versionadded` directive? You can check #15426 for examples about how this should be addressed.

cmarmo reviewed (commented):
Thanks @avigupta2612! Also, please add a new entry in the "[what's new](https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new/v0.24.rst)" file for version 0.24, describing your addition.

avigupta2612 committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into colorbar

avigupta2612 committed:
doc

avigupta2612 committed:
versionadded

cmarmo commented:
Last comments on my side.
I think that the `colorbar` parameter should be described as a boolean, this is the way you are using it in the code.
I'm not sure about the label to attach to this feature in the "what's new".

Let's wait for other reviewers.
And, please, be patient: the release is just out and people are really busy fixing all comments about the upgrade.
Thanks for your work!

amueller commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -70,6 +71,9 @@ def plot(self, *, include_values=True, cmap='viridis',
             Axes object to plot on. If `None`, a new figure and axes is
             created.
 
+        colorbar : matplotlib colorbar, default=True
```
```suggestion
        colorbar : bool, default=True
```

amueller commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -70,6 +71,9 @@ def plot(self, *, include_values=True, cmap='viridis',
             created.
 
+        colorbar : matplotlib colorbar, default=True
+            If `False` colorbar to the plot is not added.
```
Maybe say "Whether to add a colorbar to the plot"?

avigupta2612 replied:
Thanks @amueller for reviewing. I will work on those suggestions.

avigupta2612 committed:
bool

amueller reviewed (approved):
Looks good. Maybe check that by default the colorbar is not None?

avigupta2612 commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -115,8 +119,8 @@ def plot(self, *, include_values=True, cmap='viridis',
             display_labels = self.display_labels
-
-        fig.colorbar(self.im_, ax=ax)
+        if colorbar:
+            fig.colorbar(self.im_, ax=ax)
```
@amueller Here, I have added if condition to make sure colorbar is added only if parameter is true and by default its value is true. Also, I have checked the code on a working model.

venkyyuvy commented on doc/whats_new/v0.24.rst:
```diff
@@ -65,6 +65,12 @@ Changelog
+:mod:`sklearn.metrics`
+......................
+
+- |Feature| :func:`metrics.plot_confusion_matrix now supports making colorbar 
```
I think closing ` is missing for tag.

```suggestion
- |Feature| :func:`metrics.plot_confusion_matrix` now supports making colorbar 

```

avigupta2612 committed:
Update doc/whats_new/v0.24.rst

Co-authored-by: Venkatachalam N <venky.yuvy@gmail.com>

glemaitre commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -173,6 +173,9 @@ def test_confusion_matrix_display(pyplot, data, fitted_clf, y_pred, n_classes):
     assert_allclose(image_data, cm)
 
+    disp.plot(colorbar=False)
+    assert disp.im_.colorbar is None
```
I think that we should add a separate test to check the effect of the parameter.
Indeed, it is appearing in `disp.plot` and `plot_confusion_matrix` so we should check that the behaviour is correct.

```python
@pytest.mark.parametrize("colorbar, [True, False])
def test_plot_confusion_matrix_colorbar(colorbar):
    disp = plot_confusion_matrix(..., colorbar=colorbar)
    assert disp.im_.colorbar ...
    
     # attempt a plot with the opposit effect
     disp.plot(colorbar=not colorbar)
    assert disp.im_.colorbar ...
```

thomasjpfan replied:
This still needs to be addressed.

glemaitre reviewed (commented):
I think that we can improve the test mainly just to cover all possible use cases.

cmarmo commented:
Hi @avigupta2612 , you already have one approval. Do you have some time to fix conflicts and apply last comments? Thanks for your work!

cmarmo commented:
Hi @avigupta2612 I can't find a way your changes could produce a failure in the tests. Do you mind fixing conflicts and synchronizing with upstream? Thanks!

cmarmo commented:
Hi @avigupta2612 I think that something went wrong in the update: this PR is updating 131 files now, with a number of blank lines removed... making the lint fail.
The last relevant commit was 9fcc4d98619fe838b0dfbbe3e6e26d28ea445aa0, do you mind reverting to that point and re-synchronized?
You can use the following workflow:
```
$ git reset 741b3b0fdd5e0c570547ed009f2606f1659c8f2c
```
- check the status of your repo, add and commit if necessary
```
$ git fetch upstream
$ git merge upstream/master
```
- check possible conflicts.
```
$ git push --force origin colorbar
```
_You need the `--force` option to force the push otherwise git will ask you to git pull your branch before_

Hope this will help you to finalize your PR. Thanks for your patience.

avigupta2612 committed:
Merge branch 'master' into colorbar

avigupta2612 commented:
@cmarmo  Thanks for the help. I have updated the pr as you suggested. I don't know what went wrong in the previous two merges.  I hope this merge will be fine.

cmarmo commented:
> @cmarmo Thanks for the help. I have updated the pr as you suggested. I don't know what went wrong in the previous two merges. I hope this merge will be fine.

You're welcome, and I hope you will forgive me if I'm telling you that I put a typo in the workflow (the commit was 9fcc4d98619fe838b0dfbbe3e6e26d28ea445aa0)... :(
The last comments from @glemaitre went lost... Could you reapply them? Thanks for your patience!

avigupta2612 committed:
Update confusion_matrix.py

avigupta2612 commented:
@cmarmo I made the changes as per this 9fcc4d9 commit.

cmarmo commented:
@glemaitre or perhaps @thomasjpfan, all green here and one approval already. Do you mind having a look? Thanks!

thomasjpfan reviewed (commented):
Thank you for the PR @avigupta2612 !

avigupta2612 committed:
test

avigupta2612 commented:
I created a separate test for colorbar as suggestes by @glemaitre. I am running lint test and it is working fine but when I push the changes the tests are faling can you help here @cmarmo

cmarmo commented:
Some conflicts with upstream raised up, in the meanwhile. Merging with upstream (and solving conflicts), will solve your issues.

avigupta2612 committed:
conf

cmarmo commented on doc/whats_new/v0.24.rst:
```diff
@@ -230,12 +234,15 @@ Changelog
   :meth:`tree.DecisionTreeRegressor.fit`, and has not effect.
   :pr:`17614` by :user:`Juan Carlos Alfaro Jiménez <alfaro96>`.
 
+<<<<<<< HEAD
+=======
```
Those two lines should be removed: they are the relic of the automatic git diff with upstream

cmarmo commented on doc/whats_new/v0.24.rst:
```diff
@@ -230,12 +234,15 @@ Changelog
   files downloaded or cached to ensure data integrity.
   :pr:`14800` by :user:`Shashank Singh <shashanksingh28>` and `Joel Nothman`_.
 
+>>>>>>> upstream/master
```
Same here: to be removed.

cmarmo commented on doc/whats_new/v0.24.rst:
```diff
@@ -230,12 +234,15 @@ Changelog
 
+<<<<<<< HEAD
+=======
 :mod:`sklearn.datasets`
 .......................
 - |Feature| :func:`datasets.fetch_openml` now validates md5checksum of arff
   files downloaded or cached to ensure data integrity.
   :pr:`14800` by :user:`Shashank Singh <shashanksingh28>` and `Joel Nothman`_.
 
```
Indeed the entire block should be removed: it is already present on the top.
Modules are in alphabetical order in the what's new file.

avigupta2612 committed:
new

cmarmo commented:
@thomasjpfan all is green here but I did a mistake: the current what's new page contains two entries for the `sklearn.datasets` module (see [here](https://scikit-learn.org/dev/whats_new/v0.24.html)). I asked @avigupta2612 to remove the second one but it wasn't a duplicate. Can we merge this PR and I will propose a fix for my mistake? I don't have edit right and I think that a lot of syncs have already be done in this PR. Thanks for your understanding.

avigupta2612 commented:
Thanks @cmarmo for your help. The whats_new file I updated ([here](https://github.com/scikit-learn/scikit-learn/blob/2fd8603f850719e135b40dd22afa0ce60824c8d4/doc/whats_new/v0.24.rst)) and the main page  [here](https://scikit-learn.org/dev/whats_new/v0.24.html) both contain same entries for sklearn.datasets module. I guess the entry I removed in the previous commit was a third entry in sklearn.datasets

cmarmo commented:
> Thanks @cmarmo for your help. The whats_new file I updated ([here](https://github.com/scikit-learn/scikit-learn/blob/2fd8603f850719e135b40dd22afa0ce60824c8d4/doc/whats_new/v0.24.rst)) and the main page [here](https://scikit-learn.org/dev/whats_new/v0.24.html) both contain same entries for sklearn.datasets module. I guess the entry I removed in the previous commit was a third entry in sklearn.datasets

Right, if you still have some time do you mind re-adding it at the beginning of the file? Then I hope that I don't need to re-synchronize again before approval. Thanks for your patience.

avigupta2612 commented:
@cmarmo I guess this third entry for sklearn.datasets has been updated incorrectly in the main master branch [here](https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new/v0.24.rst). Should I change it accoring to the master branch or should I include it in the sklearn.datasets section?

cmarmo commented:
> @cmarmo I guess this third entry for sklearn.datasets has been updated incorrectly in the main master branch [here](https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new/v0.24.rst). Should I change it accoring to the master branch or should I include it in the sklearn.datasets section?

I think the best option is let it were it is in master and fix this in another pull request. Thanks.

> Or If I can give you the edit rights to you tell me how to do it so that you can change it accordingly.

You are very reactive and you know how to fix the issue, no need for that. :)

avigupta2612 committed:
Merge remote-tracking branch 'upstream/master' into colorbar

glemaitre commented on doc/whats_new/v0.24.rst:
```diff
@@ -230,12 +234,6 @@ Changelog
   :meth:`tree.DecisionTreeRegressor.fit`, and has not effect.
   :pr:`17614` by :user:`Juan Carlos Alfaro Jiménez <alfaro96>`.
 
-:mod:`sklearn.datasets`
```
You should reintroduce these lines (233-238) at the exact same position.

avigupta2612 replied:
I have included entry for my change at lines (145-148). So shouldn't I include skleanrn.datasets entry at lines 237-242?

glemaitre replied:
These lines are red in the diff because you removed them (probably not on purpose but when fixing a conflict). Thus, you should reinsert such that we don't see any diff on GitHub (meaning that you did not edit the lines apart of your additional entry).

glemaitre commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -246,6 +246,17 @@ def test_confusion_matrix_pipeline(pyplot, clf, data, n_classes):
+@pytest.mark.parametrize("colorbar", [True, False])
+def test_plot_confusion_matrix_colorbar(pyplot, data, fitted_clf, colorbar):
+    X, y = data
+    disp = plot_confusion_matrix(fitted_clf, X, y, colorbar=True)
+    assert disp.im_.colorbar is not None
+
+    # attempt a plot with the opposit effect
+    disp.plot(colorbar=False)
+    assert disp.im_.colorbar is None
```
```suggestion

    def _check_colorbar(disp, has_colorbar):
        if has_colorbar:
            assert disp.im_.colorbar is not None
            assert disp.im_.colorbar.__class__.__name__ == "Colorbar"
        else:
            assert disp.im_.colorbar is None

    disp = plot_confusion_matrix(fitted_clf, X, y, colorbar=colorbar)
    _check_colorbar(disp, colorbar)

    # attempt a plot with the opposite effect of colorbar
    disp.plot(colorbar=not colorbar)
    _check_colorbar(disp, not colorbar)
```

glemaitre replied:
We need to use the `colorbar` created by the parametrization.

glemaitre reviewed (commented):
2 changes and we are good to go.

avigupta2612 committed:
Merge remote-tracking branch 'upstream/master' into colorbar

avigupta2612 committed:
test

avigupta2612 commented:
@glemaitre I have made all the changes as you suggested thanks for your help.

glemaitre commented:
@avigupta2612 Thanks for the work
